drivers/volume/portworx/portworx.go

```diff
@@ -949,53 +952,201 @@ func (p *portworx) SnapshotDelete(snapDataSrc *crdv1.VolumeSnapshotDataSource, _
+			})
+			if err != nil {
+				if _, ok := err.(*ost_errors.ErrExists); !ok {
+					return err
+				}
+			}
+			volInfo := &stork_crd.RestoreVolumeInfo{}
+			volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusInProgress
+			volInfo.Volume = volID
+			volInfo.Reason = "Volume restore in progress"
+			volumeInfos = append(volumeInfos, volInfo)
+		}
+		snapRestore.Status.Volumes = volumeInfos
+		snapRestore.Status.RestoreType = string(crdv1.PortworxSnapshotTypeCloud)
+		return nil
+	default:
+		return fmt.Errorf("invalid SourceType for snapshot(local/cloud)")
+	}
+}
+
+func (p *portworx) GetVolumeSnapshotRestoreStatus(snapRestore *stork_crd.VolumeSnapshotRestore) error {
+	// Get volume client from user context
+	volDriver, err := p.getUserVolDriver(snapRestore.Annotations)
+	if err != nil {
+		return err
+	}
+
+	// Nothing to do for local snapshot
+	if snapRestore.Status.RestoreType == string(crdv1.PortworxSnapshotTypeLocal) {
+		return nil
+	}
+	for _, volInfo := range snapRestore.Status.Volumes {
+		taskID := restoreTaskPrefix + volInfo.Volume
+		csStatus := p.getCloudSnapStatus(volDriver, api.CloudRestoreOp, taskID)
+		if isCloudsnapStatusActive(csStatus.status) {
+			volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusInProgress
+			volInfo.Reason = "Volume restore in progress"
+		} else if isCloudsnapStatusFailed(csStatus.status) {
+			volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusFailed
+			volInfo.Reason = fmt.Sprintf("Restore failed for volume: %v", csStatus.msg)
+		} else {
+			// check for ha update and then mark as successful
+			isUpdate, err := p.checkAndUpdateHaLevel(volInfo.Volume, snapRestore.Annotations)
+			if err != nil {
+				volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusFailed
+				volInfo.Reason = fmt.Sprintf("Restore ha-update failed for volume: %v", err.Error())
+				return nil
+			}
+			if !isUpdate {
+				volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusResyncInProgress
+				volInfo.Reason = "Volume is in resync state"
+				return nil
+			}
+			volInfo.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusSuccessful
+			volInfo.Reason = fmt.Sprintf("Restore successful for volume")
+		}
+	}
+
+	return nil
+}
+
+func (p *portworx) checkAndUpdateHaLevel(volID string, params map[string]string) (bool, error) {
+	// Get volume client from user context
+	volDriver, err := p.getUserVolDriver(params)
+	if err != nil {
+		return false, err
+	}
+	// Get Volume Info
+	vols, err := volDriver.Inspect([]string{volID, restoreNamePrefix + volID})
+	if err != nil {
+		return false, &ErrFailedToInspectVolume{
+			ID:    volID,
+			Cause: fmt.Sprintf("Volume inspect returned err: %v", err),
+		}
+	} else if len(vols) == 0 {
+		return false, &ErrFailedToInspectVolume{
+			ID:    volID,
+			Cause: fmt.Sprintf("Volume not found err: %v", err),
+		}
+	}
+	logrus.Debugf("volume info %v \t %v", vols[0].GetLocator().GetName(), vols[1].GetLocator().GetName())
+	restoreVol := vols[1]
```

Comment:
The order of the returned volumes should not be assumed. Make sure the first is the original and the second is the restored volume